lib/improver/wind_direction.py

```diff
@@ -254,55 +317,58 @@ def calc_confidence_measure(wind_dir_complex, wind_dir_deg_mean,
                                        0.0, dist_from_mean_norm)
-        return dist_from_mean_norm
+        self.confidence_slice = self.wdir_slice_mean.copy(
+            data=dist_from_mean_norm)
+        return
 
-    @staticmethod
-    def wind_dir_decider(wind_dir_deg, wind_dir_deg_mean, r_vals, r_thresh):
+    def wind_dir_decider(self):
         """If the wind direction is so widely scattered that the r value
            is nearly zero then this indicates that the average wind direction
            is essentially meaningless.
            We therefore substitute this meaningless average wind
            direction value for the wind direction taken from the first
            ensemble realization.
 
-        Args:
-            wind_dir_deg (np.ndarray):
+        Uses:
+            self.wdir_slice.data (np.ndarray):
                 3D array - wind direction angles from ensembles (in degrees).
-            wind_dir_deg_mean (np.ndarray):
+            self.wdir_slice_mean.data (np.ndarray):
                 2D array - average wind direction angle (in degrees).
-            r_vals (np.ndarray):
+            self.r_vals_slice.data (np.ndarray):
                 2D array - Radius taken from average complex wind direction
                 angle.
-            r_thresh (float):
+            self.r_thresh (float):
                 Any r value below threshold is regarded as meaningless.
 
-        Returns:
-            reprocessed_wind_dir_mean (np.ndarray):
-                3D array - Wind direction degrees where ambigious values have
+        Defines:
+            self.wdir_slice_mean.data (np.ndarray):
+                2D array - Wind direction degrees where ambigious values have
                 been replaced with data from first ensemble realization.
         """
 
         # Mask True if r values below threshold.
-        where_low_r = np.where(r_vals < r_thresh, True, False)
+        where_low_r = np.where(self.r_vals_slice.data < self.r_thresh, True,
+                               False)
 
         # If the whole array contains good r-values, return origonal array.
         if not where_low_r.any():
-            return wind_dir_deg_mean
+            return
 
-        # Takes first ensemble realization.
-        first_realization = wind_dir_deg[0]
+        if self.backup_method == 'neighbourhood':
+            improved_values = np.full_like(self.wdir_slice_mean.data, None)
+        else:
+            # Takes first ensemble realization.
+            improved_values = self.wdir_slice.data[0]
 
         # If the r-value is low - subistite average wind direction value for
         # the wind direction taken from the first ensemble realization.
-        reprocessed_wind_dir_mean = np.where(where_low_r, first_realization,
-                                             wind_dir_deg_mean)
+        self.wdir_slice_mean.data = np.where(where_low_r, improved_values,
+                                             self.wdir_slice_mean.data)
+        return
```

Comment:
I don't like this (multiple null "return" statements) but I can see why you've done it.

Thinking about how to factor this differently: any reason why you couldn't check for low r-values outside the function, and call it conditionally?  That tidies things up a little.  (But this is nitpicky, so I'll say it's a "nice to have".)

Reply:
Agreed. I'll make this change.